registry/registry-polaris/src/main/java/com/alipay/sofa/rpc/registry/polaris/PolarisRegistry.java

```diff
@@ -0,0 +1,369 @@
+            }
+            return;
+        }
+        if (!config.isRegister()) {
+            return;
+        }
+
+        try {
+            List<InstanceRegisterRequest> instanceRegisterRequests = buildPolarisRegister(config);
+
+            for (InstanceRegisterRequest request : instanceRegisterRequests) {
+                deregisterPolarisService(config, request);
+            }
+
+        } catch (Exception e) {
+            if (!RpcRunningState.isShuttingDown()) {
+                if (e instanceof SofaRpcRuntimeException) {
+                    throw e;
+                } else {
+                    throw new SofaRpcRuntimeException(LogCodes.getLog(LogCodes.ERROR_UNREG_PROVIDER, EXT_NAME), e);
+                }
+            }
+        }
+    }
+
+    private void deregisterPolarisService(ProviderConfig config, InstanceRegisterRequest request) {
+
+        InstanceDeregisterRequest instanceDeregisterRequest = new InstanceDeregisterRequest();
+        instanceDeregisterRequest.setNamespace(request.getNamespace());
+        instanceDeregisterRequest.setService(request.getService());
+        instanceDeregisterRequest.setHost(request.getHost());
+        instanceDeregisterRequest.setPort(request.getPort());
+        providerAPI.deRegister(instanceDeregisterRequest);
+
+        ScheduledFuture future = heartbeatFutures.remove(buildUniqueName(config, request.getProtocol()));
+        if (future != null) {
+            future.cancel(true);
+        }
+    }
+
+    @Override
+    public void batchUnRegister(List<ProviderConfig> configs) {
+        for (ProviderConfig config : configs) {
+            unRegister(config);
+        }
+    }
+
+    @Override
+    public List<ProviderGroup> subscribe(ConsumerConfig config) {
+        String appName = config.getAppName();
+        if (!registryConfig.isSubscribe()) {
+            if (LOGGER.isInfoEnabled(appName)) {
+                LOGGER.infoWithApp(appName, LogCodes.getLog(LogCodes.INFO_REGISTRY_IGNORE));
+            }
+            return null;
+        }
+        if (!config.isSubscribe()) {
+            return null;
+        }
+
+        try {
+            List<ProviderInfo> providers = findService(config);
+            if (EventBus.isEnable(ConsumerSubEvent.class)) {
+                ConsumerSubEvent event = new ConsumerSubEvent(config);
+                EventBus.post(event);
+            }
+
+            return Collections.singletonList(new ProviderGroup().addAll(providers));
+        } catch (SofaRpcRuntimeException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new SofaRpcRuntimeException(LogCodes.getLog(LogCodes.ERROR_SUB_PROVIDER, EXT_NAME), e);
+        }
+
+    }
+
+    private List<ProviderInfo> findService(ConsumerConfig config) {
+
+        String uniqueName = buildUniqueName(config, config.getProtocol());
+        PolarisWatcher watcher = polarisWatchers.get(uniqueName);
```

Comment:
It is not thread safe. Consumers with same uniqueId may register to different watcher, and only one watch will keep in memory. So some consumer's listener may not receive data any more.

Reply:
@OrezzerO  fixed and commited